File: specter/tests/unit/checker/check.test.ts
/**
 * Tests for spec-check: the type checker.
 *
 * @spec spec-check
 */

import { describe, it, expect } from 'vitest';
import { checkSpecs } from '../../../src/core/checker/check.js';
import {
  classifyChanges,
  highestClassification,
} from '../../../src/core/checker/rules/breaking-changes.js';
import type { SpecGraph, SpecNode } from '../../../src/core/resolver/types.js';
import type { SpecAST } from '../../../src/core/schema/types.js';

function makeSpec(overrides: Partial<SpecAST> & { id: string }): SpecAST {
  return {
    version: '1.0.0',
    status: 'approved',
    tier: 2,
    context: { system: 'test' },
    objective: { summary: 'test' },
    constraints: [{ id: 'C-01', description: 'test constraint' }],
    acceptance_criteria: [
      { id: 'AC-01', description: 'test ac', references_constraints: ['C-01'] },
    ],
    ...overrides,
  };
}

function makeGraph(
  specs: Array<{ spec: SpecAST; file: string }>,
  edges: SpecGraph['edges'] = [],
): SpecGraph {
  const nodes = new Map<string, SpecNode>();
  for (const s of specs) {
    nodes.set(s.spec.id, s);
  }
  return {
    nodes,
    edges,
    topological_order: specs.map((s) => s.spec.id),
    diagnostics: [],
  };
}

describe('spec-check', () => {
  // @ac AC-01: Spec with constraint C-03 not referenced by any AC produces OrphanConstraint
  // References: C-01
  it('AC-01: detects orphan constraint not referenced by any AC', () => {
    const spec = makeSpec({
      id: 'test-orphan',
      constraints: [
        { id: 'C-01', description: 'referenced' },
        { id: 'C-02', description: 'referenced' },
        { id: 'C-03', description: 'NOT referenced' },
      ],
      acceptance_criteria: [
        { id: 'AC-01', description: 'test', references_constraints: ['C-01'] },
        { id: 'AC-02', description: 'test', references_constraints: ['C-02'] },
      ],
    });

    const graph = makeGraph([{ spec, file: 'test.spec.yaml' }]);
    const result = checkSpecs(graph);

    const orphans = result.diagnostics.filter((d) => d.kind === 'orphan_constraint');
    expect(orphans).toHaveLength(1);
    expect(orphans[0].constraint_id).toBe('C-03');
    expect(orphans[0].spec_id).toBe('test-orphan');
  });

  // @ac AC-02: Tier 1 orphan is error; Tier 3 is info
  // References: C-01, C-02
  it('AC-02: respects tier-based severity for orphan constraints', () => {
    const tier1Spec = makeSpec({
      id: 'tier1-spec',
      tier: 1,
      constraints: [{ id: 'C-01', description: 'orphan' }],
      acceptance_criteria: [{ id: 'AC-01', description: 'test' }],
    });
    const tier3Spec = makeSpec({
      id: 'tier3-spec',
      tier: 3,
      constraints: [{ id: 'C-01', description: 'orphan' }],
      acceptance_criteria: [{ id: 'AC-01', description: 'test' }],
    });

    const graph1 = makeGraph([{ spec: tier1Spec, file: 't1.spec.yaml' }]);
    const graph3 = makeGraph([{ spec: tier3Spec, file: 't3.spec.yaml' }]);

    const result1 = checkSpecs(graph1);
    const result3 = checkSpecs(graph3);

    const orphan1 = result1.diagnostics.find((d) => d.kind === 'orphan_constraint');
    const orphan3 = result3.diagnostics.find((d) => d.kind === 'orphan_constraint');

    expect(orphan1).toBeDefined();
    expect(orphan1!.severity).toBe('error');
    expect(orphan3).toBeDefined();
    expect(orphan3!.severity).toBe('info');
  });

  // @ac AC-03: Structural conflict between upstream required and downstream absent
  // References: C-03
  it('AC-03: detects structural conflict between connected specs', () => {
    const upstream = makeSpec({
      id: 'user-registration',
      constraints: [{ id: 'C-01', description: 'email MUST be required' }],
      acceptance_criteria: [{ id: 'AC-01', description: 'test', references_constraints: ['C-01'] }],
    });
    const downstream = makeSpec({
      id: 'guest-checkout',
      depends_on: [{ spec_id: 'user-registration', relationship: 'requires' }],
      constraints: [{ id: 'C-01', description: 'test' }],
      acceptance_criteria: [
        {
          id: 'AC-01',
          description: 'Process checkout when email is absent',
          references_constraints: ['C-01'],
        },
      ],
    });

    const graph = makeGraph(
      [
        { spec: upstream, file: 'upstream.spec.yaml' },
        { spec: downstream, file: 'downstream.spec.yaml' },
      ],
      [{ from: 'guest-checkout', to: 'user-registration', relationship: 'requires' }],
    );

    const result = checkSpecs(graph);
    const conflicts = result.diagnostics.filter((d) => d.kind === 'structural_conflict');

    expect(conflicts.length).toBeGreaterThanOrEqual(1);
    expect(conflicts[0].severity).toBe('error');
  });

  // @ac AC-04: Removing a constraint is classified as breaking (MAJOR)
  // References: C-04
  it('AC-04: classifies removed constraint as breaking change', () => {
    const v1 = makeSpec({
      id: 'test',
      constraints: [
        { id: 'C-01', description: 'keep' },
        { id: 'C-02', description: 'will be removed' },
      ],
      acceptance_criteria: [
        { id: 'AC-01', description: 'test', references_constraints: ['C-01'] },
        { id: 'AC-02', description: 'test', references_constraints: ['C-02'] },
      ],
    });
    const v2 = makeSpec({
      id: 'test',
      constraints: [{ id: 'C-01', description: 'keep' }],
      acceptance_criteria: [{ id: 'AC-01', description: 'test', references_constraints: ['C-01'] }],
    });

    const changes = classifyChanges(v1, v2);
    expect(highestClassification(changes)).toBe('breaking');

    const removals = changes.filter((c) => c.classification === 'breaking');
    expect(removals.length).toBeGreaterThanOrEqual(1);
  });

  // @ac AC-05: Adding an optional field is classified as additive (MINOR)
  // References: C-04
  it('AC-05: classifies added AC as additive change', () => {
    const v1 = makeSpec({
      id: 'test',
      constraints: [{ id: 'C-01', description: 'test' }],
      acceptance_criteria: [
        { id: 'AC-01', description: 'original', references_constraints: ['C-01'] },
      ],
    });
    const v2 = makeSpec({
      id: 'test',
      constraints: [{ id: 'C-01', description: 'test' }],
      acceptance_criteria: [
        { id: 'AC-01', description: 'original', references_constraints: ['C-01'] },
        { id: 'AC-02', description: 'new addition', references_constraints: ['C-01'] },
      ],
    });

    const changes = classifyChanges(v1, v2);
    expect(highestClassification(changes)).toBe('additive');
  });

  // @ac AC-06: All constraints referenced = zero orphan diagnostics
  // References: C-05
  it('AC-06: produces zero orphan diagnostics when all constraints referenced', () => {
    const spec = makeSpec({
      id: 'fully-covered',
      constraints: [
        { id: 'C-01', description: 'a' },
        { id: 'C-02', description: 'b' },
      ],
      acceptance_criteria: [
        { id: 'AC-01', description: 'test', references_constraints: ['C-01'] },
        { id: 'AC-02', description: 'test', references_constraints: ['C-02'] },
      ],
    });

    const graph = makeGraph([{ spec, file: 'test.spec.yaml' }]);
    const result = checkSpecs(graph);

    const orphans = result.diagnostics.filter((d) => d.kind === 'orphan_constraint');
    expect(orphans).toHaveLength(0);
  });

  // Breaking change detection via checkSpecs with previousVersions
  it('reports breaking changes through checkSpecs orchestrator', () => {
    const current = makeSpec({
      id: 'evolving',
      constraints: [{ id: 'C-01', description: 'kept' }],
      acceptance_criteria: [{ id: 'AC-01', description: 'test', references_constraints: ['C-01'] }],
    });
    const previous = makeSpec({
      id: 'evolving',
      constraints: [
        { id: 'C-01', description: 'kept' },
        { id: 'C-02', description: 'removed' },
      ],
      acceptance_criteria: [
        { id: 'AC-01', description: 'test', references_constraints: ['C-01'] },
        { id: 'AC-02', description: 'test', references_constraints: ['C-02'] },
      ],
    });

    const graph = makeGraph([{ spec: current, file: 'test.spec.yaml' }]);
    const result = checkSpecs(graph, {
      previousVersions: new Map([['evolving', previous]]),
    });

    const breaking = result.diagnostics.filter((d) => d.kind === 'breaking_change');
    expect(breaking.length).toBeGreaterThanOrEqual(1);
    expect(result.summary.errors).toBeGreaterThan(0);
  });
});

File: specter/internal/checker/check_test.go
// @spec spec-check
package checker

import (
	"testing"

	"github.com/Hanalyx/specter/internal/resolver"
	"github.com/Hanalyx/specter/internal/schema"
)

func makeSpec(id string, tier int) schema.SpecAST {
	return schema.SpecAST{
		ID: id, Version: "1.0.0", Status: "approved", Tier: tier,
		Context:   schema.SpecContext{System: "test"},
		Objective: schema.SpecObjective{Summary: "test"},
		Constraints: []schema.Constraint{
			{ID: "C-01", Description: "test constraint"},
		},
		AcceptanceCriteria: []schema.AcceptanceCriterion{
			{ID: "AC-01", Description: "test ac", ReferencesConstraints: []string{"C-01"}},
		},
	}
}

func makeGraph(nodes map[string]*resolver.SpecNode, edges []resolver.SpecEdge) *resolver.SpecGraph {
	return &resolver.SpecGraph{Nodes: nodes, Edges: edges}
}

// @ac AC-01
func TestOrphanConstraint(t *testing.T) {
	spec := makeSpec("test", 2)
	spec.Constraints = append(spec.Constraints,
		schema.Constraint{ID: "C-02", Description: "referenced"},
		schema.Constraint{ID: "C-03", Description: "NOT referenced"},
	)
	spec.AcceptanceCriteria = append(spec.AcceptanceCriteria,
		schema.AcceptanceCriterion{ID: "AC-02", Description: "test", ReferencesConstraints: []string{"C-02"}},
	)

	g := makeGraph(map[string]*resolver.SpecNode{"test": {Spec: spec, File: "test.yaml"}}, nil)
	result := CheckSpecs(g, nil)

	orphans := 0
	for _, d := range result.Diagnostics {
		if d.Kind == "orphan_constraint" && d.ConstraintID == "C-03" {
			orphans++
		}
	}
	if orphans != 1 {
		t.Errorf("expected 1 orphan for C-03, got %d", orphans)
	}
}

// @ac AC-02
func TestTierBasedSeverity(t *testing.T) {
	tier1 := makeSpec("t1", 1)
	tier1.Constraints = []schema.Constraint{{ID: "C-01", Description: "orphan"}}
	tier1.AcceptanceCriteria = []schema.AcceptanceCriterion{{ID: "AC-01", Description: "test"}}

	tier3 := makeSpec("t3", 3)
	tier3.Constraints = []schema.Constraint{{ID: "C-01", Description: "orphan"}}
	tier3.AcceptanceCriteria = []schema.AcceptanceCriterion{{ID: "AC-01", Description: "test"}}

	g1 := makeGraph(map[string]*resolver.SpecNode{"t1": {Spec: tier1, File: "t1.yaml"}}, nil)
	g3 := makeGraph(map[string]*resolver.SpecNode{"t3": {Spec: tier3, File: "t3.yaml"}}, nil)

	r1 := CheckSpecs(g1, nil)
	r3 := CheckSpecs(g3, nil)

	var sev1, sev3 string
	for _, d := range r1.Diagnostics {
		if d.Kind == "orphan_constraint" {
			sev1 = d.Severity
		}
	}
	for _, d := range r3.Diagnostics {
		if d.Kind == "orphan_constraint" {
			sev3 = d.Severity
		}
	}

	if sev1 != "error" {
		t.Errorf("expected tier 1 orphan severity 'error', got %q", sev1)
	}
	if sev3 != "info" {
		t.Errorf("expected tier 3 orphan severity 'info', got %q", sev3)
	}
}

// @ac AC-03
func TestStructuralConflict(t *testing.T) {
	upstream := makeSpec("user-reg", 1)
	upstream.Constraints = []schema.Constraint{{ID: "C-01", Description: "email MUST be required"}}
	upstream.AcceptanceCriteria = []schema.AcceptanceCriterion{{ID: "AC-01", Description: "test", ReferencesConstraints: []string{"C-01"}}}

	downstream := makeSpec("guest", 2)
	downstream.Constraints = []schema.Constraint{{ID: "C-01", Description: "test"}}
	downstream.AcceptanceCriteria = []schema.AcceptanceCriterion{
		{ID: "AC-01", Description: "Process checkout when email is absent", ReferencesConstraints: []string{"C-01"}},
	}

	g := makeGraph(
		map[string]*resolver.SpecNode{
			"user-reg": {Spec: upstream, File: "u.yaml"},
			"guest":    {Spec: downstream, File: "g.yaml"},
		},
		[]resolver.SpecEdge{{From: "guest", To: "user-reg", Relationship: "requires"}},
	)

	result := CheckSpecs(g, nil)
	found := false
	for _, d := range result.Diagnostics {
		if d.Kind == "structural_conflict" {
			found = true
		}
	}
	if !found {
		t.Error("expected structural_conflict diagnostic")
	}
}

// @ac AC-04
func TestBreakingChangeRemoval(t *testing.T) {
	v1 := makeSpec("test", 2)
	v1.Constraints = []schema.Constraint{
		{ID: "C-01", Description: "keep"},
		{ID: "C-02", Description: "removed"},
	}
	v1.AcceptanceCriteria = []schema.AcceptanceCriterion{
		{ID: "AC-01", Description: "test", ReferencesConstraints: []string{"C-01"}},
		{ID: "AC-02", Description: "test", ReferencesConstraints: []string{"C-02"}},
	}

	v2 := makeSpec("test", 2)

	changes := ClassifyChanges(&v1, &v2)
	if HighestClassification(changes) != "breaking" {
		t.Errorf("expected breaking, got %s", HighestClassification(changes))
	}
}

// @ac AC-05
func TestAdditiveChange(t *testing.T) {
	v1 := makeSpec("test", 2)
	v2 := makeSpec("test", 2)
	v2.AcceptanceCriteria = append(v2.AcceptanceCriteria,
		schema.AcceptanceCriterion{ID: "AC-02", Description: "new"},
	)

	changes := ClassifyChanges(&v1, &v2)
	if HighestClassification(changes) != "additive" {
		t.Errorf("expected additive, got %s", HighestClassification(changes))
	}
}

// @ac AC-06
func TestNoOrphansWhenAllReferenced(t *testing.T) {
	spec := makeSpec("full", 1)
	spec.Constraints = []schema.Constraint{
		{ID: "C-01", Description: "a"},
		{ID: "C-02", Description: "b"},
	}
	spec.AcceptanceCriteria = []schema.AcceptanceCriterion{
		{ID: "AC-01", Description: "test", ReferencesConstraints: []string{"C-01"}},
		{ID: "AC-02", Description: "test", ReferencesConstraints: []string{"C-02"}},
	}

	g := makeGraph(map[string]*resolver.SpecNode{"full": {Spec: spec, File: "f.yaml"}}, nil)
	result := CheckSpecs(g, nil)

	for _, d := range result.Diagnostics {
		if d.Kind == "orphan_constraint" {
			t.Errorf("unexpected orphan: %s", d.ConstraintID)
		}
	}
}

// @ac AC-07
func TestStrictModeUpgradesWarningsToErrors(t *testing.T) {
	// Tier 2 orphan is normally a warning
	spec := makeSpec("mid", 2)
	spec.Constraints = []schema.Constraint{
		{ID: "C-01", Description: "referenced"},
		{ID: "C-02", Description: "orphan"},
	}
	spec.AcceptanceCriteria = []schema.AcceptanceCriterion{
		{ID: "AC-01", Description: "test", ReferencesConstraints: []string{"C-01"}},
	}

	g := makeGraph(map[string]*resolver.SpecNode{"mid": {Spec: spec, File: "mid.yaml"}}, nil)

	// Without strict: warning
	r := CheckSpecs(g, nil)
	for _, d := range r.Diagnostics {
		if d.Kind == "orphan_constraint" && d.Severity != "warning" {
			t.Errorf("expected warning without strict, got %q", d.Severity)
		}
	}
	if r.Summary.Errors > 0 {
		t.Error("expected no errors without strict mode")
	}

	// With strict: error
	rs := CheckSpecs(g, &CheckOptions{Strict: true})
	for _, d := range rs.Diagnostics {
		if d.Kind == "orphan_constraint" && d.Severity != "error" {
			t.Errorf("expected error with strict=true, got %q", d.Severity)
		}
	}
	if rs.Summary.Errors == 0 {
		t.Error("expected errors > 0 in strict mode")
	}
}

// @ac AC-08
func TestWarnOnDraftEmitsDraftSpecDiagnostic(t *testing.T) {
	spec := makeSpec("draft-spec", 2)
	spec.Status = "draft"

	g := makeGraph(map[string]*resolver.SpecNode{"draft-spec": {Spec: spec, File: "d.yaml"}}, nil)

	// Without warn_on_draft: no draft diagnostic
	r := CheckSpecs(g, nil)
	for _, d := range r.Diagnostics {
		if d.Kind == "draft_spec" {
			t.Error("unexpected draft_spec diagnostic without WarnOnDraft")
		}
	}

	// With warn_on_draft: warning emitted
	rw := CheckSpecs(g, &CheckOptions{WarnOnDraft: true})
	found := false
	for _, d := range rw.Diagnostics {
		if d.Kind == "draft_spec" && d.SpecID == "draft-spec" && d.Severity == "warning" {
			found = true
		}
	}
	if !found {
		t.Error("expected draft_spec warning diagnostic with WarnOnDraft=true")
	}
}
